libcxx/src/time_zone.cpp

```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
+   * time marked with ~~ is not a valid local time. This is represented by the
+   * nonexistent value in local_info.result.
+   *
+   * At the first of November the time goes backward one hour. This means the
+   * time marked with ^^ happens twice. This is represented by the ambiguous
+   * value in local_info.result.
+   *
+   * 2020.11.01                  2021.04.01              2021.11.01
+   * offset +05                  offset +05              offset +05
+   * save    0s                  save    1h              save    0s
+   * |-------------W----------|
+   *                             |----------W--------------|
+   *                                                    |-------------
+   *                           ~~                        ^^
+   *
+   * These shifts can happen due to changes in the current time zone for a
+   * location. For example, Indian/Kerguelen switched only once. In 1950 from an
+   * offset of 0 hours to an offset of +05 hours.
+   *
+   * During all these shifts the UTC time will have not gaps.
+   */
+
+  // The code needs to determine the system time for the local time. There is no
+  // information available. Assume the offset between system time and local time
+  // is 0s. This gives an initial estimate.
+  sys_seconds __guess{__local_seconds};
+  sys_info __info = __get_info(__guess);
+
+  // At this point the offset can be used to determine an estimate for the local
+  // time. Before doing the determine the offset validate whether the local time
+  // is the range [chrono::local_seconds::min(), chrono::local_seconds::max()).
+  if (__local_seconds < 0s && __info.offset > 0s)
+    if (__local_seconds - chrono::local_seconds::min().time_since_epoch() < __info.offset)
+      return {-1, __info, {}};
+
+  if (__local_seconds > 0s && __info.offset < 0s)
+    if (chrono::local_seconds::max().time_since_epoch() - __local_seconds < -__info.offset)
+      return {-2, __info, {}};
+
+  // Based on the information in the sys_info found the local time can be
+  // converted to a system time. This resulting time can be in the following
+  // locations of the sys_info found:
+  //
+  //                             |----------W--------------|
+  //                           1   2        3             4  5
+  //
+  // 1. The estimate is before the returned sys_info object.
+  //    The result is either non-existent or unique in the previous sys_info.
+  // 2. The estimate is in the beginning of the returned sys_info object.
+  //    The result is either unique or ambiguous with the previous sys_info.
+  // 3. The estimate is in the "middle" of the returned sys_info.
+  //    The result is unique.
+  // 4. The result is at the end of the returned sys_info object.
+  //    The result is either unique or ambiguous with the next sys_info.
+  // 5. The estimate is after the returned sys_info object.
+  //    The result is either non-existent or unique in the next sys_info.
```

Comment:
The notion of "at the beginning of a sys info object" and "at the end of [...]" seems a bit arbitrary to me. Naively, I would instead say that we are either "within" a sys info object, or not. I would basically:

1. Is the time before the "current" sys_info object? If so, non-existent or unique in the previous one.
2. Is the time within the "current" sys_info object? If so, check whether that time is also in the previous one (in which case it's ambiguous), or if that time is also in the next one (ambiguous too). If not, that time is uniquely within the "current" sys_info.
3. Is the time after the "current" sys_info? If so, non-existent or unique in the next.

Doing it this way would be more expensive cause we have to materialize the previous/next sys_infos, but I think it's the correct algorithm to check what we want here. And optimizations to materializing sys_infos can be done to make this more efficient (such optimizations would be useful on their own anyway).

Reply:
I'm not too fond of your proposed wording, I've used different wording.
The implementation now no longer detects a "middle". It now tests the previous and next sys_info object.